crates/sui-types/src/coin.rs

```diff
@@ -109,6 +116,20 @@ impl Coin {
             ],
         }
     }
+
+    pub fn merge_coins(&mut self, coins_to_merge: &mut [Coin]) {
+        let total_coins = coins_to_merge.iter().fold(0, |acc, c| acc + c.value());
```

Comment:
I think a coin should implement a merge function that merges one coin (pass by value) into another (by mut ref) with combined balance. Then this can be called for all coins to merge all balances. This is important to make sure that you only merge a coin ever once.
In the current implementation, it's a bit strange that you are emptying each coin without actually destroying them. (you wouldn't need to explicitly withdraw all value with an unwrap)
This is a nit and can be done in follow up PRs.

Reply:
> it's a bit strange that you are emptying each coin without actually destroying them

agreed while I think this is due to the fact that, deletion is about temp store and `coin_object: Object`, and this function is about `coin: Coin`. Technically we can pass in temp store and let this function to do the coin_object deletion as well but I think it's a bit cleaner to not involve temp store in`impl Coin`. same thing for split_coin, we create a new `Coin` and not write into temp store in that function.

Just added a couple of comments to `merge_coins` and `split_coin`, and will resolve comments in a following PR if needed, thanks a lot!